### ISSUE TITLE
TouchEnd event is fired before finger leaves the screen when dragging is disabled

### ISSUE THREAD

**Issue by zorji** [Leaflet/Leaflet#5745]
## How to reproduce

- Leaflet version I'm using: 1.0.3/1.2.0
- Browser (with version) I'm using: Chrome 60.0.3112.107
- OS/Platform (with version) I'm using: Android 7.1.2

```
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Leaflet</title>
    <link rel="stylesheet" href="https://unpkg.com/leaflet@1.2.0/dist/leaflet.css"
          integrity="sha512-M2wvCLH6DSRazYeZRIm1JnYyh22purTM+FDB5CsyxtQJYeKq83arPe5wgbNmcFXGqiSH2XR8dT/fJISVA1r/zQ=="
          crossorigin=""/>
    <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no"/>
    <style type="text/css">
        body {
            padding: 0;
            margin: 0;
        }

        html, body, #map {
            height: 100vh;
            width: 100vw;
        }
    </style>
</head>
<body>
<div id="map" style="height: 500px;"></div>
<p>Line 1</p>
<p>Line 1</p>
<p>Line 1</p>
<p>Line 1</p>
<p>Line 5</p>
<p>Line 1</p>
<p>Line 1</p>
<p>Line 1</p>
<p>Line 5</p>
<p>Line 1</p>
<p>Line 1</p>
<p>Line 1</p>
<p>Line 1</p>
<p>Line 5</p>
<p>Line 1</p>
<p>Line 1</p>
<p>Line 1</p>
<script src="https://unpkg.com/leaflet@1.2.0/dist/leaflet.js"
        integrity="sha512-lInM/apFSqyy1o6s89K4iQUKg6ppXEgsVxT35HbzUupEVRh2Eu9Wdl4tHj7dZO0s1uvplcYGmt3498TtHq+log=="
        crossorigin=""></script>
<script>
    (function () {
        var handler = L.Handler.extend({
            addHooks: function () {
                L.DomEvent.on(this._map._container, 'touchstart', function (e) {
                    console.log('leaflet - touchstart');
                }, this);
                L.DomEvent.on(this._map._container, 'touchmove', function (e) {
                    console.log('leaflet - touchmove');
                }, this);
                L.DomEvent.on(this._map._container, 'touchend', function (e) {
                    console.log('leaflet - touchend');
                }, this);
            }
        });

        L.Map.addInitHook('addHandler', 'disableSingleFingerDraggingOnTouchDevices', handler);
    })(L);
</script>
<script>
    (function (L) {
        var map = L.map('map', {
            dragging: false,
            disableSingleFingerDraggingOnTouchDevices: true
        }).setView([51.505, -0.09], 13);
        L.tileLayer('https://api.tiles.mapbox.com/v4/{id}/{z}/{x}/{y}.png?access_token=pk.eyJ1IjoibWFwYm94IiwiYSI6ImNpejY4NXVycTA2emYycXBndHRqcmZ3N3gifQ.rJcFIG214AriISLbB6B5aw', {
            maxZoom: 18,
            attribution: 'Imagery © <a href="http://mapbox.com">Mapbox</a>',
            id: 'mapbox.streets'
        }).addTo(map);
    })(L);

    window.addEventListener('touchstart', function () {
        console.log('native - touchstart');
    });
    window.addEventListener('touchend', function () {
        console.log('native - touchend');
    });
    window.addEventListener('touchmove', function () {
        console.log('native - touchmove');
    });
</script>
</body>
</html>
```

## What behaviour I'm expecting and which behaviour I'm seeing

Try to drag on the map, I expect the touchend event from Leaflet fires when my fingers leave the screen. However, the touchend event fires while I am still moving my touch. The native touch event fire correctly. This is working fine in iOS Safari. I haven't tested Chrome in iOS.

## Minimal example reproducing the issue

- [*] this example is as simple as possible
- [*] this example does not rely on any third party code

Example http://playground-leaflet.rhcloud.com/virac/edit?html,output

**Comment by perliedman** [comment]
I can reproduce this on Chrome 62/Linux with this plunkr: https://plnkr.co/edit/uY3jC8Q24Q605tVMLajN?p=preview

On my laptop with touch screen, swiping with the finger across the map produces this console output:

```
leaflet - touchstart
script.js:32 native - touchstart
5 script.js:8 leaflet - touchmove
script.js:38 native - touchmove
script.js:8 leaflet - touchmove
script.js:38 native - touchmove
script.js:11 leaflet - touchend
155 script.js:38 native - touchmove
script.js:35 native - touchend
```
(The number in front of some of the events are Chrome's repeat count).

As can be seen, Leaflet produces a few `touchmove` events, but almost immediately fires `touchend`, while the native touch events continue a long time.

**Comment by moklick** [comment]
I am having the same problem. The only way I can "fix" this is using `leaflet@1.0.0-beta.1`. The error seems to start with version 1.0.0.

**Comment by hjrobinson** [comment]
moklick were you able to find a working solution with latest version of leaflet? I'll do some experimenting with cutting a pasting between the two versions but I thought I would check first to see if there was an existing solution.  I would rather avoid using the older version because it has some bugs that showed up immediately.

**Comment by johnd0e** [comment]
Ok, I've tracked down the root of the issue: recent leaflet instead of listening for native touch events - relies on their simulation based on pointer events, see yourself:

https://jsbin.com/rulimax/edit?html,output

And pointer events sometimes have flow different from touch: for not-draggable targets 'pointercancel' is produced early.
<details>

Details of events behavior can be observed here: https://rbyers.github.io/eventTest.html
</details>

BTW, in Firefox I often see 'touchend' without corresponding 'pointerup'/'pointercancel', which can be a problem too.

---
Proposed fix: #7029/3.

**Comment by johnd0e** [comment]
As appeared, similar issue is reported in #5969.
And alternative fix can be here:
https://github.com/Leaflet/Leaflet/blob/d843c3b88486713827d7e860b58bdba75bfbd5a2/dist/leaflet.css#L58-L61

Which need to be replaced with `touch-action: none;`
